// @w11k/ngx-componentdestroyed/dist/tslint/BadComponent.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var index_1 = require("../index");
var SomeClass = /** @class */ (function () {
    function SomeClass() {
    }
    SomeClass.prototype.subscribe = function () {
    };
    return SomeClass;
}());
var BadComponent = /** @class */ (function () {
    function BadComponent() {
        this.observable = rxjs_1.of(1);
        this.stop = rxjs_1.of(1);
    }
    BadComponent.prototype.ngOnInit = function () {
        // Error
        this.observable.pipe(operators_1.map(function (i) { return i; }), operators_1.takeUntil(this.stop), operators_1.map(function (i) { return i; })).subscribe();
        // Error
        this.observable.pipe(operators_1.map(function (i) { return i; }), index_1.untilComponentDestroyed(this), operators_1.map(function (i) { return i; })).subscribe();
        // OK
        this.observable.pipe(operators_1.map(function (i) { return i; }), operators_1.takeUntil(this.stop)).subscribe();
        // OK
        this.observable.pipe(operators_1.map(function (i) { return i; }), index_1.untilComponentDestroyed(this)).subscribe();
        // OK
        new SomeClass().subscribe();
    };
    BadComponent.prototype.ngOnDestroy = function () {
    };
    BadComponent = __decorate([
        core_1.Component({})
    ], BadComponent);
    return BadComponent;
}());
